File: part-0/python/primitive_test_framework.py
class PrimitiveTestFramework:
    def __init__(self):
        self.passed_tests = 0
        self.failed_tests = 0
        self.test_number = 0

    def assert_equal(self, name, actual, expected):
        self.test_number += 1
        if actual == expected:
            self.passed_tests += 1
            self.print_one_result(name, True)
        else:
            self.failed_tests += 1
            self.print_one_result(
                name, False, f"Expected {expected}, Actual: {actual}")

    def assert_true(self, name, expression):
        self.assert_equal(name, expression, True)

    def end_tests(self):
        print("--- Test results: ---")
        if self.failed_tests == 0:
            print("✅ All tests passed!")
        else:
            print(f"✅: {self.passed_tests}, ❌: {self.failed_tests}")

    def print_one_result(self, name: str, passed: bool, message: str | None = None):
        print(
            f"{'✅' if passed else '❌'} [{self.test_number}] {name}")
        if message:
            print("     | " + message)

File: part-0/java-solution/PrimitiveTestFramework.java
public class PrimitiveTestFramework {
    private int passedTests;
    private int failedTests;
    private int testNumber;

    public PrimitiveTestFramework() {
        passedTests = 0;
        failedTests = 0;
        testNumber = 0;
    }

    public <T> void assertEqual(
            String name,
            T actual,
            T expected) {
        testNumber++;
        if (actual.equals(expected)) {
            passedTests++;
            printOneResult(name, true, "");
        } else {
            failedTests++;
            printOneResult(name, false, "Expected: " + expected + ", Actual: " + actual);
        }
    }

    public void assertTrue(
            String name,
            boolean expression) {
        assertEqual(name, expression, true);
    }

    public void endTests() {
        System.out.println("--- Test results: ---");
        if (failedTests == 0)
            System.out.println("✅ All tests passed!");
        else
            System.out.println("✅: " + passedTests + ", ❌: " + failedTests);
    }

    private void printOneResult(
            String name,
            boolean passed,
            String message) {
        System.out.println((passed ? "✅" : "❌") + " [" + testNumber + "] " + name);
        if (message != null && message.length() > 0)
            System.out.println("    | " + message);
    }

}